bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/provider/ScriptModuleTypeProvider.java

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.core.automation.module.script.internal.provider;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameter;
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameter.Type;
+import org.eclipse.smarthome.config.core.ConfigDescriptionParameterBuilder;
+import org.eclipse.smarthome.config.core.ParameterOption;
+import org.eclipse.smarthome.core.common.registry.ProviderChangeListener;
+import org.openhab.core.automation.Visibility;
+import org.openhab.core.automation.module.script.ScriptEngineFactory;
+import org.openhab.core.automation.module.script.internal.handler.AbstractScriptModuleHandler;
+import org.openhab.core.automation.module.script.internal.handler.ScriptActionHandler;
+import org.openhab.core.automation.module.script.internal.handler.ScriptConditionHandler;
+import org.openhab.core.automation.type.ActionType;
+import org.openhab.core.automation.type.ConditionType;
+import org.openhab.core.automation.type.ModuleType;
+import org.openhab.core.automation.type.ModuleTypeProvider;
+import org.openhab.core.automation.type.Output;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class dynamically provides ScriptActionType and ScriptConditionType {@link ModuleType}s. This class is necessary
+ * because there is no other way to provide dynamic {@link ParameterOption}s for {@link ModuleType}s.
+ *
+ * @author Scott Rushworth - Initial contribution
+ */
+@Component
+public class ScriptModuleTypeProvider implements ModuleTypeProvider {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final List<ParameterOption> parameterOptions = new CopyOnWriteArrayList<>();
+
+    @SuppressWarnings("unchecked")
```

Comment:
What is unchecked here?

Reply:
Lol... I was wondering the same thing earlier today! I copied it over in the code I copied from [MediaActionModuleTypeProvider](https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.automation.module.media/src/main/java/org/openhab/core/automation/module/media/internal/MediaActionTypeProvider.java#L53).

Reply:
Type safety: The return type ModuleType for getModuleType(String, Locale) from the type ScriptModuleTypeProvider needs unchecked conversion to conform to T from the 
	 type ModuleTypeProvider

Reply:
IMHO the API of `ModuleTypeProvider` needs to be changed. It does not make sense to use a unbound return value.
It will be unknown in any case...

We should improve that API.